drivers/random/random_nrf5.c

```diff
@@ -0,0 +1,116 @@
+/*
+ * Copyright (c) 2017 Exati Tecnologia Ltda.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <random.h>
+#include <atomic.h>
+#include <nrf_rng.h>
+
+struct random_nrf5_dev_data {
+	atomic_t user_count;
+};
+
+#define DEV_DATA(dev) \
+	((struct random_nrf5_dev_data *)(dev)->driver_data)
+
+static inline s16_t random_nrf5_get_u8(void)
+{
+	u8_t value;
+	u32_t timeout = 0;
+
+	while (!nrf_rng_event_get(NRF_RNG_EVENT_VALRDY)) {
+		timeout++;
+		if (timeout >= 1000000) {
+			return -ETIMEDOUT;
+		}
+	}
+
+	value = nrf_rng_random_value_get();
+	nrf_rng_event_clear(NRF_RNG_EVENT_VALRDY);
+
+	return value;
+}
+
+static int random_nrf5_get_entropy(struct device *device, u8_t *buf, u16_t len)
+{
+	int ret = 0;
+
+	/* Mark the peripheral as being used */
+	atomic_inc(&DEV_DATA(device)->user_count);
+
+	/* Disable the shortcut that stops the task after a byte is generated */
+	nrf_rng_shorts_disable(NRF_RNG_SHORT_VALRDY_STOP_MASK);
+
+	/* Start the RNG generator peripheral */
+	nrf_rng_task_trigger(NRF_RNG_TASK_START);
+
+	while (len) {
+		s16_t value = random_nrf5_get_u8();
+
+		if (value < 0) {
+			ret = value;
+			break;
+		}
+
+		*buf = (u8_t) value;
+		buf++;
+		len--;
+	}
+
+	/* Only stop the RNG generator peripheral if we're the last user */
+	if (atomic_dec(&DEV_DATA(device)->user_count) == 1) {
+		/* Disable the peripheral on the next VALRDY event */
+		nrf_rng_shorts_enable(NRF_RNG_SHORT_VALRDY_STOP_MASK);
+
+		if (atomic_get(&DEV_DATA(device)->user_count) != 0) {
+			/* Race condition: another thread started to use
+			 * the peripheral while we were disabling it.
+			 * Enable the peripheral again
+			 */
+			nrf_rng_shorts_disable(NRF_RNG_SHORT_VALRDY_STOP_MASK);
+			nrf_rng_task_trigger(NRF_RNG_TASK_STOP);
```

Comment:
`NRF_RNG_TASK_START`

Reply:
Wow -- sorry about that and thanks for the catch. Sometimes it's best to go to sleep instead of programming, but that's the time I have to contribute. Fixed